Incubating AvroTransformer in SwissKnife

eoterogarcia opened the pull request:
Incubating AvroTransformer in SwissKnife

eoterogarcia committed:
Incubating AvroTransformer in SwissKnife

ju4nlu commented on CHANGELOG.md:
```diff
@@ -6,6 +6,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+## [0.6.0] - 2020-05-05
```
Be careful with this! Version number and release date should be added right before merging the PR, not when creating it. A new PR may come in and be merged before this one, so this version and release date would not be correct

eoterogarcia replied:
Big true

ju4nlu commented on SwissKnife/avro/AvroTransformer.py:
```diff
@@ -0,0 +1,271 @@
+Record = Dict[str, object]
+
+
+class NoDefault(object):
```
What's this class used for? A little description would be nice

ju4nlu commented on SwissKnife/avro/AvroTransformer.py:
```diff
@@ -0,0 +1,271 @@
+        renamed_record = self.get_renamed_record_and_remove_invalid_fields(record)
+        transformed_record = self.get_transformed_record(renamed_record)
+        record_with_defaults = self.get_record_with_defaults(transformed_record)
+        record_with_casted_values = self.get_record_with_casted_values(record_with_defaults)
```
I'd just return in this line, instead of creating variable `record_with_casted_values`

ju4nlu commented on SwissKnife/avro/AvroTransformer.py:
```diff
@@ -0,0 +1,271 @@
+        """Transforms a record no a new one where empty field will
+           be filled with its default value.
+           If a field is empty and has not default value,
+           the function will raise an exception.
```
I'd rewrite this doc with:
> Given a record, creates a new one where empty fields are filled with its default value, defined in the loaded schema. If an empty field does not have a default value, then an exception will arise.

ju4nlu commented on SwissKnife/avro/AvroTransformer.py:
```diff
@@ -0,0 +1,271 @@
+           the function will raise an exception.
+        :param record: The input record.
+        :type record: Record
+        :raises RuntimeError: When a field empty and has not default value.
```
literature in this section
> When there is an empty field with no default value

ju4nlu commented on SwissKnife/avro/AvroTransformer.py:
```diff
@@ -0,0 +1,271 @@
+        new_record = {}
+        for key, defaults_value in self.defaults_dict.items():
+            if key not in record and defaults_value is NoDefault:
+                raise RuntimeError(f"Required  {key} not in record {record}")
```
Two spaces after `Required` :P

ju4nlu commented on SwissKnife/avro/AvroTransformer.py:
```diff
@@ -0,0 +1,271 @@
+        """
+
+        rename_dict = {}
+        for field in schema["fields"]:
```
It seems like it might be a good idea to move this strings which match the keys of the schema ("fields", "name", "aliases", ....) into a class variable. What do you think?

eoterogarcia replied:
Good idea, very convenient!

ju4nlu commented on SwissKnife/avro/AvroTransformer.py:
```diff
@@ -0,0 +1,271 @@
+        return transform_dict
+
+    @staticmethod
+    def _create_cast_dict(schema: dict) -> dict:
```
Missing docs!

ju4nlu commented on SwissKnife/avro/AvroTransformer.py:
```diff
@@ -0,0 +1,271 @@
+
+    @staticmethod
+    def _get_transform_function(transform_name: str) -> Callable[[object, dict], object]:
+        """Gets a transform function from a transform_name. This purpose of this function
```
Change "This purpose of this function is transform" to "`The` purpose of this function is `to` transform"

ju4nlu commented on SwissKnife/avro/AvroTransformer.py:
```diff
@@ -0,0 +1,271 @@
+        is transform the current value of a field to another one (for example, an available
+        value for the field type).
+        transform_name can be directly the name of a function or it can be
+        a expresion, so the result will be a closure. For example,
```
I'd rewrite this sentence to something like:
> transform_name could be the name of a function or an expression, so the result will always be a closure.

ju4nlu commented on SwissKnife/avro/AvroTransformer.py:
```diff
@@ -0,0 +1,271 @@
+            - record: The entire record.
+
+        Available functions:
+            - int2boolean: Transform a int to a bolean. True if x > 0, otherwise False.
```
Transform `an` int to a `boolean`

ju4nlu commented on SwissKnife/avro/AvroTransformer.py:
```diff
@@ -0,0 +1,271 @@
+                if copied_field in record:
+                    return record[copied_field]
+                else:
+                    raise RuntimeError("Invalid field in copyFrom")
```
Maybe the exception message could contain more info to help debug the issue, such as the value of `copied_field` or the available fields in the record

ju4nlu commented on SwissKnife/avro/AvroTransformer.py:
```diff
@@ -0,0 +1,271 @@
+            selected_function = copyFrom
+        # If there is not an available function, raise Exception
+        if selected_function is None:
+            raise RuntimeError("Invalid name for a transform function")
```
Same here, I'd print the name that was given

ju4nlu commented on SwissKnife/avro/AvroTransformer.py:
```diff
@@ -0,0 +1,271 @@
+                if type(value) is str and len(value) == 0:
+                    return None
+                else:
+                    return int(value)
```
Could we have problems when casting really big `long` values? I guess it's not a problem we are facing now nor should correct right now, I was just wondering...

eoterogarcia replied:
According to [python documentation](https://docs.python.org/3.6/library/stdtypes.html#typesnumeric): 

> There are three distinct numeric types: integers, floating point numbers, and complex numbers. [...] **Integers have unlimited precision.** [...]

So in this case, I think is not an issue.

ju4nlu replied:
cool! thanks for the link :+1:

ju4nlu commented on SwissKnife/avro/AvroTransformer.py:
```diff
@@ -0,0 +1,271 @@
+    def _get_casted_value(key: str, value: object, cast_dict: dict) -> object:
+        types_list = cast_dict[key]
+        for type_to_cast in types_list:
+            if type_to_cast == "null" and value is None:
```
Should this condition be an `OR`? If not, I can see that some `None` values will filter to other if statements and they might cause an error

eoterogarcia replied:
I think you are right. Testing the code, it fail casting to int and float, while casting to boolean produce a false value and casting to string, a 'None' string.
I'll change it, but just to be sure, what do you think, @nradz ?

eoterogarcia replied:
It is something behind, because changing the if statement to `OR` breaks test `test_apply_all_transforms`. I'll research what happens

ju4nlu commented on tests/avro/test_AvroTransformer.py:
```diff
@@ -0,0 +1,170 @@
+            }
+        )
+
+    def test_get_record_with_defaults(self):
```
When there is an empty record, is it also completely filled with defaults?

eoterogarcia replied:
No, it isn't. It rise an error with the message `Required url not in record {}` 
In this example test, _url_ is a field with no default value, so it is automatically consider as a required field (also _date_)

ju4nlu replied:
Ok, thanks for the explanation!

ju4nlu reviewed (changes requested):
There are somethings that should be changed (mostly docs) and I've asked some questions in certain parts of the code. Overall I think it's quite good, but I guess @nradz can provide a more in-depth review of the technical aspects of this implementation.

eoterogarcia committed:
Move changes to unreleased in CHANGELOG.md

eoterogarcia committed:
Return value instead of creating a new variable

eoterogarcia committed:
Fix documentation in get_record_with_defaults function

eoterogarcia committed:
Fix space typo

eoterogarcia committed:
Add Variables class

Add Variables class with static reference to string values used in code, also moved Record to types.py

eoterogarcia committed:
Minor fixes

Change condition in if statement inside _get_casted_value(). Include TODO comments to resolve in later commits

eoterogarcia committed:
Rollback the if condition change

Minor changes in test

eoterogarcia committed:
Improvement in documentation

New function description were provided. Also a minor change in code was made to improve error message.

eoterogarcia committed:
Change deprecated methods

Change assertEquals to assertEqual

eoterogarcia committed:
NoDefault class documentation

eoterogarcia committed:
Move changes to a released item in CHANGELOG.md

eoterogarcia committed:
Merge branch 'master' into feature/incubating_avrotransformer

# Conflicts:
#	CHANGELOG.md
